MODEL_NO_TOOLS_USED error since v4.145.0 in VSCode Kilo Code plugin

### Plugin Type

VSCode Extension

### App Version

v4.149.0

### Description

Since v4.146.0 I've been seeing this error for every and any request in VSCode Kilo Code plugin.
Checked each version up to v4.149.0.

v4.145.0 and earlier versions don't have this issue.
Doesn't matter if I tell Kilocode to read a file or just ask "What day is it today?"

Tried reinstalling the Kilo Code plugin and clearing data. No effect. 

Found only two immediately effective workarounds:
A) Downgrading to v4.145.0. Nano-GPT models work again. OK as a temporary solution.
B) Changing to Kilo Code provider, Grok fast model (free). Works, but not a solution unfortunately.

Basic error message:
Date/time: 2026-01-19T23:07:48.902Z
Extension version: 4.149.0
Provider: nano-gpt
Model: zai-org/glm-4.7:thinking
MODEL_NO_TOOLS_USED

Related JSON log excerpt:
{
      "role": "user",
      "content": [
        {
          "type": "text",
          "text": "[ERROR] You did not use a tool in your previous response! Please retry with a tool use.
...

### Reproduction steps

1. Install VSCode(v1.108.1) Kilo Code plugin v4.146.0-v4.149.0
2. Choose API provider: Nano-GPT. Model List: Subscription Models. 
3. Model: zai-org/glm-4.7, zai-org/glm-4.7:thinking or zai-org/glm-4.6 or what looks like any other model from this provider.
4. Make any request in any mode. Like "What day is it today?" or "Read [filename]. Do you see any issues?"

### Provider

Nano-GPT

### Model

 zai-org/glm-4.7:thinking

### System Information

Laptop: Lenovo Legion 5
OS: Nobara linux latest
CPU: Ryzen 7 5800H with Radeon Graphics, 8 Cores.
GPU: NVIDIA GeForce RTX 3050 Ti Laptop
RAM: 16GB
VSCode v1.108.1

--- a/.changeset/fix-model-no-tools-used.md
+++ b/.changeset/fix-model-no-tools-used.md
@@ -1,0 +1,5 @@
+---
+"kilo-code": patch
+---
+
+Fix recurring MODEL_NO_TOOLS_USED error loop by detecting text-based tool call hallucinations and instructing the model to use the native API.

--- a/src/core/prompts/sections/__tests__/tool-use-guidelines.spec.ts
+++ b/src/core/prompts/sections/__tests__/tool-use-guidelines.spec.ts
@@ -54,9 +54,8 @@ describe("getToolUseGuidelinesSection", () => {
 				expect(guidelines).toContain("1. Assess what information")
 				expect(guidelines).toContain("2. Choose the most appropriate tool")
 				expect(guidelines).toContain("3. If multiple actions are needed")
-				expect(guidelines).toContain("4. After each tool use")
+				expect(guidelines).toContain("5. After each tool use")
 			})
-
 			it("should include single-tool-per-message guidance when experiment disabled", () => {
 				const guidelines = getToolUseGuidelinesSection(TOOL_PROTOCOL.NATIVE, {})
 

--- a/src/core/prompts/sections/tool-use-guidelines.ts
+++ b/src/core/prompts/sections/tool-use-guidelines.ts
@@ -37,6 +37,10 @@ export function getToolUseGuidelinesSection(
 				`${itemNumber++}. If multiple actions are needed, use one tool at a time per message to accomplish the task iteratively, with each tool use being informed by the result of the previous tool use. Do not assume the outcome of any tool use. Each step must be informed by the previous step's result.`,
 			)
 		}
+
+		guidelinesList.push(
+			`${itemNumber++}. CRITICAL: You must use the API's native tool format. Do NOT simply write text describing the tool use (e.g., "[Tool Use: ...]" or JSON blocks in text). The system will strictly reject any text that mimics a tool call. You must use the proper API structure for function calling.`,
+		)
 	} else {
 		guidelinesList.push(
 			`${itemNumber++}. If multiple actions are needed, use one tool at a time per message to accomplish the task iteratively, with each tool use being informed by the result of the previous tool use. Do not assume the outcome of any tool use. Each step must be informed by the previous step's result.`,

--- a/src/core/task/Task.ts
+++ b/src/core/task/Task.ts
@@ -3607,6 +3607,11 @@ export class Task extends EventEmitter<TaskEvents> implements TaskLike {
 					)
 
 					if (!didToolUse) {
+						// Check for hallucinated tool use pattern
+						const hallucinatedTool = this.assistantMessageContent.find(
+							(block) => block.type === "text" && block.content.trim().match(/^\[Tool Use: .+\]/i),
+						)
+
 						// Increment consecutive no-tool-use counter
 						this.consecutiveNoToolUseCount++
 
@@ -3617,10 +3622,17 @@ export class Task extends EventEmitter<TaskEvents> implements TaskLike {
 							this.consecutiveMistakeCount++
 						}
 
+						let responseText = formatResponse.noToolsUsed(this._taskToolProtocol ?? "xml")
+
+						if (hallucinatedTool) {
+							responseText +=
+								"\n\n[ERROR] You are outputting tool calls as text (e.g. '[Tool Use: ...]'). This is invalid. You MUST use the native tool calling capability provided by the API. Do not write the tool use in the text response."
+						}
+
 						// Use the task's locked protocol for consistent behavior
 						this.userMessageContent.push({
 							type: "text",
-							text: formatResponse.noToolsUsed(this._taskToolProtocol ?? "xml"),
+							text: responseText,
 						})
 					} else {
 						// Reset counter when tools are used successfully

